### PR TITLE
BucketListDB Bucket Apply Optimization

### LINKED ISSUE #4113: Improve bucket apply
The Bucket Apply step is currently very slow. We iterate through the entire BucketList in reverse order, leading to significant disk write churn. When BucketListDB is enabled, we only need to apply offers. Instead of reading the entire BucketList in reverse order, we should only scan the section of the BucketList that contains offers in-order when BucketsDB is enabled.

### PR DESCRIPTION
**Pull request by SirTyson**
# Description

Resolves #4113

This change improves Bucket apply times by approximately 80% when BucketListDB is enabled. In the current catchup flow, we first apply buckets then index those buckets. The Bucket apply step scans the entire BucketList even though only offers need to be applied. This PR changes the ordering such that we index buckets then apply them. This allows us to use the index to only scan sections of the BucketList that contain offers.

Bucket apply currently writes every entry it sees to the SQL DB. Because the BucketList is implemented via a LSMT, many versions of a given entry may exist in the BucketList, with only the most recent version being valid. This means Bucket apply must iterate through the BucketList in reverse order such that the last version of a key written is the most recent. This causes significant write amplification, as many versions of the same entry are written to disk even though only the most recent version is valid.

When BucketListDB is enabled, this change now applies buckets in-order so that we must only write any key a single time. However, to avoid writing out of date entries to the DB, in-order application requires that every key written to the DB is recorded in an in-memory set. This set would be too large if all entry types are being applied. However, when BucketListDB is enabled, we must only apply offer entries, making this optimization possible. 

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/bucket/BucketIndexImpl.cpp [hunk 1]
```diff
@@ -450,6 +443,39 @@ BucketIndexImpl<IndexT>::getPoolshareTrustlineRange(
...
+    // Get the smallest and largest possible offer keys
+    LedgerKey upperBound(OFFER);
+    upperBound.offer().sellerID.ed25519().fill(
+        std::numeric_limits<uint8_t>::max());
+    upperBound.offer().offerID = std::numeric_limits<int64_t>::max();
+
+    LedgerKey lowerBound(OFFER);
+    lowerBound.offer().sellerID.ed25519().fill(
+        std::numeric_limits<uint8_t>::min());
+    lowerBound.offer().offerID = std::numeric_limits<int64_t>::min();
```
Comment (sisuresh) [inline src/bucket/BucketIndexImpl.cpp:474]: The minimum possible `offerID` is 0, not `INT64_MIN`, but this doesn't matter in practice because `getOffsetBounds` does a binary search using this `lowerBound`. Is my understanding correct?

Comment (SirTyson) [reply src/bucket/BucketIndexImpl.cpp:474]: Correct, we just want the smallest possible dummy OFFER key so that we can properly seek the the first place in the file where the OFFER region starts. Whether or not the dummy key we use is actually valid doesn't matter.

**Review by graydon — APPROVED** [review]
LGTM!

File: src/catchup/ApplyBucketsWork.cpp [hunk 2]
```diff
@@ -174,31 +230,54 @@ ApplyBucketsWork::startLevel()
     auto& level = getBucketLevel(mLevel);
     HistoryStateBucket const& i = mApplyState.currentBuckets.at(mLevel);
 
-    bool applySnap = (i.snap != binToHex(level.getSnap()->getHash()));
-    bool applyCurr = (i.curr != binToHex(level.getCurr()->getHash()));
+    bool applyFirst = mOffersOnly
```
Comment (marta-lokhova) [inline src/catchup/ApplyBucketsWork.cpp:233]: all the new ternary operators are getting a bit hard to follow; that aside, this function applying the whole level has always been kind of strange. like, is there actually a reason to apply the whole level and keep track of snap & curr? Alternatively, we could just have an iterator of buckets in the correct order of application, and `doWork` could pick up one bucket at a time to process. This way I think we don't need to track snap and curr at all. This does require quite a bit of refactoring though, so maybe we can at least open an issue to clean this up?

**Review by marta-lokhova — CHANGES_REQUESTED** [review]
Curious to see catchup application time after this change!

File: src/bucket/BucketApplicator.h [hunk 3]
```diff
@@ -28,6 +28,10 @@ class BucketApplicator
     BucketInputIterator mBucketIter;
     size_t mCount{0};
     std::function<bool(LedgerEntryType)> mEntryTypeFilter;
+    bool const mNewOffersOnly;
+    UnorderedSet<LedgerKey>& mSeenKeys;
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/bucket/BucketApplicator.h:32]: I don't think so. This is just storing seen offer keys, so I think any kind of attack triggering lots of hash collisions would have to be very sophisticated and pretty worthless, as the result would be a marginally slowed catch up.

File: src/catchup/ApplyBucketsWork.cpp [hunk 4]
```diff
@@ -333,13 +441,15 @@ ApplyBucketsWork::advance(std::string const& bucketName,
...
+    return !(mApplying) || !(mFirstBucketApplicator || mSecondBucketApplicator);
 }
 
 std::string
 ApplyBucketsWork::getStatus() const
 {
-    if (!mSpawnedAssumeStateWork)
+    // This status string only applies to step 2 when we actually apply the
+    // buckets.
+    if (mFinishedIndexBucketsWork && !mSpawnedAssumeStateWork)
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/catchup/ApplyBucketsWork.cpp:452]: I don't think we need to, as `IndexBucketsWork` already prints logs regarding indexing progress.

File: src/bucket/BucketApplicator.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/bucket/BucketApplicator.h:32]: so.. how about std::unordered_set then? :)

File: src/catchup/ApplyBucketsWork.cpp (hunk 2 above)
Comment (SirTyson) [reply src/catchup/ApplyBucketsWork.cpp:233]: Yeah, I agree that this is worth doing but should happen post merge. I've opened an issue for it here https://github.com/stellar/stellar-core/issues/4290.

File: src/catchup/ApplyBucketsWork.cpp [hunk 5]
```diff
@@ -107,9 +134,19 @@ ApplyBucketsWork::doReset()
...
+    mBucketsToIndex.clear();
 
     if (!isAborting())
     {
+        if (mOffersOnly)
+        {
+            // The current size of this set is 1.6 million during BucketApply
+            // (as of 12/20/23). There's not a great way to estimate this, so
+            // reserving with some extra wiggle room
+            mSeenKeys.reserve(2'000'000);
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/catchup/ApplyBucketsWork.cpp:147]: It's pretty variable and correlated to high DEX activity rather than BucketList size. Currently, it takes about 180 mb. I also tested a period of very high DEX activity in 2018 which consumed 1.15 GB. For replaying all history, I would estimate the upper bound is around 1.2 GB (I found an expensive period, but I'm not sure if it's the absolute max).

[the remark this answers is not in the document]
Comment (SirTyson) [reply src/catchup/ApplyBucketsWork.cpp:147]: Looks like the peak memory usage of core on any given range is 3.8 GB during catchup (total, not just the offer apply index).

File: src/catchup/ApplyBucketsWork.cpp [hunk 6]
```diff
@@ -208,8 +290,27 @@ ApplyBucketsWork::doWork()
 {
     ZoneScoped;
 
-    // Step 1: apply buckets. Step 2: assume state
-    if (!mSpawnedAssumeStateWork)
+    // Step 1: index buckets. Step 2: apply buckets. Step 3: assume state
+    bool isUsingBucketListDB = mApp.getConfig().isUsingBucketListDB();
+    if (isUsingBucketListDB)
```
Comment (marta-lokhova) [inline src/catchup/ApplyBucketsWork.cpp:295]: nit: I'd prefer if we could write this flow such that it's consistent with Work usage across the codebase:
```
if (!mIndexBucketsWork)
{
     // Spawn indexing work for the first time
     mIndexBucketsWork = addWork<IndexBucketsWork>(mBucketsToIndex);
     return State::WORK_RUNNING;
}
else if (mIndexBucketsWork->getState() != BasicWork::State::WORK_SUCCESS)
{
      // Exit early if indexing work is still running, or failed
      return mIndexBucketsWork->getState();
}

// Otherwise, continue with next steps
```

File: src/catchup/ApplyBucketsWork.cpp [hunk 7]
```diff
@@ -208,8 +290,27 @@ ApplyBucketsWork::doWork()
...
+    {
+        if (!mIndexBucketsWork)
+        {
+            mIndexBucketsWork = addWork<IndexBucketsWork>(mBucketsToIndex);
+        }
+
+        // If indexing has not finished or finished and failed, return result
+        // status. If indexing finished and succeeded, move on and spawn assume
+        // state work.
+        auto status = checkChildrenStatus();
```
Comment (marta-lokhova) [inline src/catchup/ApplyBucketsWork.cpp:305]: Please don't use `checkChildrenStatus` to determine the state of `mIndexBucketsWork`, as this parent work might have other children, which would cause the function to return unexpected results. To query state of each individual work, use that work's `getState` method.

**Review by marta-lokhova — APPROVED** [review]
thanks for the updates, LGTM! Just a minor code cleanup request. Please squash commits also.

**Comment by SirTyson** [comment]
> thanks for the updates, LGTM! Just a minor code cleanup request. Please squash commits also.

Done!

**Comment by marta-lokhova** [comment]
r+ 004a4ba9630b0c81856a9857779a014a95afa3fd
